design/storage-class-device-set.md

```diff
@@ -0,0 +1,438 @@
+    * This may be either one Job per PVC or one Job per batch of PVCs. One per
+    PVC sounds cleaner.
+1. Create an OSD Deployment for each unused but initialized PVC
+
+### Additional considerations for local storage
+
+This design can also be applied to non-cloud environments. To take advantage of
+this, the admin should deploy the
+[sig-storage-local-static-provisioner](https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner)
+to create local PVs for the desired local storage devices and then follow the
+recommended directions for [local
+PVs](https://kubernetes.io/blog/2019/04/04/kubernetes-1.14-local-persistent-volumes-ga/).
+
+### Possible Implementation: DriveGroups
+
+Creating real-world OSDs is complex:
+
+* Some configurations deploy multiple OSDs on a single drive
+* Some configurations are using more than one drive for a single OSD.
+* Deployments often look similar on multiple hosts.
+* There are some advanced configurations possible, like encrypted drives.
+
+All of these setups are valid real-world configurations that need to be
+supported.
+
+The Ceph project defines a data structure that allows defining groups of drives
+to be provisioned in a specific way by ceph-volume: Drive Groups. Drive Groups
+were originally designed to be ephemeral, but it turns out that orchestrators
+like DeepSea store them permanently in order to have a source of truth when
+(re-)provisioning OSDs. Also, Drive Groups were originally designed to be host
+specific. But the concept of hosts is not really required for the Drive Group
+data structure make sense, as they only select a subset of a set of aviailble
+drives.
+
+DeepSea has a documentation of [some example drive
+groups](https://github.com/SUSE/DeepSea/wiki/Drive-Groups#example-drive-group-files).
+A complete specification is documented in the [ceph
+documentation](http://docs.ceph.com/docs/master/mgr/orchestrator_modules/#orchestrator.DriveGroupSpec).
+
+
+#### DeviceSet vs DriveGroup
+
+A DeviceSet to provision 8 OSDs on 8 drives could look like:
+
+```yaml
+name: "my_device_set"
+count: 8
+```
+
+The Drive Group would look like so:
+
+```yaml
+host_pattern: "hostname1"
+data_devices:
+  count: 8
+```
+
+A Drive Group with 8 OSDs using a shared fast drive could look similar to this:
+
+```yaml
+host_pattern: "hostname1"
+data_devices:
+  count: 8
+  model: MC-55-44-XZ
+db_devices:
+  model: NVME
+db_slots: 8
+```
+#### ResourceRequirements and Placement
+
+Drive Groups don't yet provide orchestrator specific extensions, like resource
+requirements or placement specs, but that could be added trivially. Also a name
+could be added to Drive Groups.
+
+### OSD Configuration Examples
+
+Given the complexity of this design, here are a few examples to showcase
+possible configurations for OSD `StorageClassDeviceSets`.
+
+#### Example 1: AWS cross-AZ
+
+```yaml
+type: CephCluster
+name: cluster1
+...
+spec:
+  ...
+  storage:
+    ...
+    deviceSets:
```

Comment:
Are we not renaming these to `storageClassDeviceSets`?

Reply:
Good catch.